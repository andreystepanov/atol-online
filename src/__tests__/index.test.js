require('dotenv').config()
import Atol from '..'
import moment from 'moment'
import uuid from 'uuid/v4'

const timestampWithoutSecondsFormat = 'DD.MM.YYYY HH:mm'
const timestampFormat = `${timestampWithoutSecondsFormat}:ss`

const options = {
  email: 'support@domain.ru',
  url: 'https://domain.ru',
  callback_url: 'https://domain.ru/callback',
  vat: 18,
  livemode: false,
}

const payload = {
  id: 'order_id',
  customer: {
    email: 'andreystepanov.blog@gmail.com',
    phone: '+380939777959',
  },
  items: [
    {
      type: 'service',
      name: 'Товар 1',
      price: 100.0,
      quantity: 2,
      unit_label: 'шт.',
      payment: null,
    },
    {
      type: 'service',
      name: 'Товар 2',
      price: 50.0,
      unit_label: 'шт.',
      payment: null,
    },
  ],
  // agent: {
  //   phones: [],
  //   name: '', // для money_transfer_operator
  //   address: '', // для money_transfer_operator
  // }
  // payment_type: 'online',
}

const correctionPayload = {
  id: 'correction_id',
  amount: 100.0,
  number: 'base_number',
  description: 'Описание коррекции',
  instruction: true,
  payment_type: 'advance',
}

const generatePayload = data => ({
  ...(data || payload),
  id: uuid(),
})

const correction = () => generatePayload(correctionPayload)

const log = jest.fn()

let atol

// const instance = new Atol({})

beforeEach(() => {
  atol = new Atol(options)
})

test('Atol exists', () => {
  expect(Atol).toBeDefined()
})

test('Receives all the options', () => {
  expect(atol).toMatchSnapshot()
})

describe('authenticate', () => {
  test('exists', async () => {
    expect(atol.authenticate).toBeDefined()
    expect(typeof atol.authenticate).toBe('function')
  })

  test.skip('send valid params', async () => {
    const result = await atol.authenticate()
  })
})

describe('timestamp', () => {
  test('timestamp', async () => {
    expect(Atol.timestamp).toBeDefined()
    expect(typeof Atol.timestamp).toBe('function')
  })

  test('returns formatted date from input', () => {
    const date = moment().subtract(1, 'year')
    const timestamp = Atol.timestamp(date.format())

    expect(timestamp).toEqual(date.format(timestampFormat))
  })

  test('returns default formatted date', () => {
    const timestamp = Atol.timestamp()

    expect(
      moment(timestamp, timestampFormat).format(timestampWithoutSecondsFormat),
    ).toEqual(moment().format(timestampWithoutSecondsFormat))
  })
})

describe('preparePurchaseOrRefund', () => {
  test('exists', async () => {
    expect(atol.preparePurchaseOrRefund).toBeDefined()
    expect(typeof atol.preparePurchaseOrRefund).toBe('function')
  })

  test.skip('throws error if id not provided', async () => {
    const result = await atol
      .preparePurchaseOrRefund({
        ...payload,
        id: undefined,
      })
      .catch(error => error)
  })

  test('returns prepared data', async () => {
    const { timestamp, ...result } = await atol.preparePurchaseOrRefund(payload)

    expect(timestamp).toBeDefined()
    expect(result).toMatchSnapshot()
  })

  test('sets custom timestamp and callback_url', async () => {
    const timestamp = moment('2017-11-28T23:49:45+07:00').format(
      timestampFormat,
    )
    const callbackUrl = `${options.callback_url}/new`

    const result = await atol.preparePurchaseOrRefund({
      ...payload,
      callback_url: callbackUrl,
      timestamp,
      timestamp_format: timestampFormat,
    })

    expect(result.timestamp).toEqual(timestamp)
    expect(result).toMatchSnapshot()
  })
})

describe('prepareCorrection', () => {
  const timestamp = moment('2017-11-28T23:49:45+07:00').format(timestampFormat)

  test('exists', async () => {
    expect(atol.prepareCorrection).toBeDefined()
    expect(typeof atol.prepareCorrection).toBe('function')
  })

  test('returns prepared data', async () => {
    const { timestamp, ...result } = await atol.prepareCorrection(
      correctionPayload,
    )
    const { base_date: baseDate } = result.correction.correction_info
    result.correction.correction_info.base_date = '[sanitized]'

    expect(timestamp).toBeDefined()
    expect(baseDate).toBeDefined()
    expect(result).toMatchSnapshot()
  })

  test('sets custom timestamp and callback_url', async () => {
    const timestamp = moment('2017-11-28T23:49:45+07:00').format(
      timestampFormat,
    )
    const callbackUrl = `${options.callback_url}/new`

    const result = await atol.prepareCorrection({
      ...correctionPayload,
      callback_url: callbackUrl,
      timestamp,
      timestamp_format: timestampFormat,
    })

    expect(result.timestamp).toEqual(timestamp)
    expect(result).toMatchSnapshot()
  })
})

describe.skip(
  'sending real request to API',
  () => {
    describe('sell and report', () => {
      let externalId
      let sellId

      test('returns validation error', async () => {
        const { success, response } = await atol.sell
          .create({})
          .catch(error => error)

        expect(success).toBe(false)
        expect(response.status).toBe(400)
        expect(response.body.status).toMatchSnapshot('fail')
        expect(response.body.error.text).toMatchSnapshot()
      })

      test('returns successful response', async () => {
        const data = generatePayload()
        const { success, response } = await atol.sell
          .create(data)
          .catch(error => error)

        expect(success).toBe(true)

        expect(response.status).toBe(200)
        expect(response.body.uuid).toBeDefined()
        expect(response.body.status).toBeDefined()
        expect(response.body.error).toBeNull()
        expect(response.body.timestamp).toBeDefined()

        externalId = data.id
        sellId = response.body.uuid
      })

      test('returns data about sell by uuid', async () => {
        const result = await atol.report.get(sellId).catch(error => error)
        const {
          success,
          response: { body },
        } = result

        expect(success).toBe(true)

        expect(body.uuid).toEqual(sellId)
        expect(body.external_id).toEqual(externalId)

        expect(body.callback_url).toEqual(atol.callbackUrl)
        expect(body.group_code).toEqual(atol.group)
      })
    })

    describe('refund', () => {
      test('returns validation error', async () => {
        const { success, response } = await atol.sell
          .refund({})
          .catch(error => error)

        expect(success).toBe(false)
        expect(response.status).toBe(400)
        expect(response.body.status).toMatchSnapshot('fail')
        expect(response.body.error.text).toMatchSnapshot()
      })

      test('returns successful response', async () => {
        const data = generatePayload()
        const { success, response } = await atol.sell
          .refund(data)
          .catch(error => error)

        expect(success).toBe(true)

        expect(response.status).toBe(200)
        expect(response.body.uuid).toBeDefined()
        expect(response.body.status).toBeDefined()
        expect(response.body.error).toBeNull()
        expect(response.body.timestamp).toBeDefined()
      })
    })

    describe('correction', () => {
      test('returns validation error', async () => {
        const { success, response } = await atol.sell
          .correction({})
          .catch(error => error)

        expect(success).toBe(false)
        expect(response.status).toBe(400)
        expect(response.body.status).toMatchSnapshot('fail')
        expect(response.body.error.text).toMatchSnapshot()
      })

      test('returns successful response', async () => {
        const data = correction()
        const { success, response } = await atol.sell
          .correction(data)
          .catch(error => error)

        expect(success).toBe(true)

        expect(response.status).toBe(200)
        expect(response.body.uuid).toBeDefined()
        expect(response.body.status).toBeDefined()
        expect(response.body.error).toBeNull()
        expect(response.body.timestamp).toBeDefined()
      })
    })
  },
  30 * 1000,
)
