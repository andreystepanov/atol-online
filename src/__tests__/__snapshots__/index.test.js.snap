// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Receives all the options 1`] = `
Atol {
  "agentType": null,
  "apiToken": null,
  "apiUrl": "https://testonline.atol.ru/possystem/v4",
  "apiVersion": "v4",
  "callbackUrl": "https://domain.ru/callback",
  "company": Object {
    "email": "support@domain.ru",
    "inn": "5544332219",
    "payment_address": "https://domain.ru",
  },
  "group": "v4-online-atol-ru_4179",
  "login": "v4-online-atol-ru",
  "password": "iGFFuihss",
  "paymentTypes": Object {
    "advance": 2,
    "credit": 3,
    "custom": 5,
    "custom_2": 6,
    "custom_3": 7,
    "custom_4": 8,
    "custom_5": 9,
    "online": 1,
    "other": 4,
  },
  "report": Object {
    "get": [Function],
  },
  "sell": Object {
    "correction": [Function],
    "create": [Function],
    "refund": [Function],
  },
  "userAgent": "Atol-Online-Node/0.1",
  "vatType": "vat18",
}
`;

exports[`prepareCorrection returns prepared data 1`] = `
Object {
  "correction": Object {
    "company": Object {
      "email": "support@domain.ru",
      "inn": "5544332219",
      "payment_address": "https://domain.ru",
    },
    "correction_info": Object {
      "base_date": "[sanitized]",
      "base_name": "Описание коррекции",
      "base_number": "base_number",
      "type": "instruction",
    },
    "payments": Array [
      Object {
        "sum": 100,
        "type": 2,
      },
    ],
    "vats": Array [
      Object {
        "sum": 18,
        "type": "vat18",
      },
    ],
  },
  "external_id": "correction_id",
  "service": Object {
    "callback_url": "https://domain.ru/callback",
  },
}
`;

exports[`prepareCorrection sets custom timestamp and callback_url 1`] = `
Object {
  "correction": Object {
    "company": Object {
      "email": "support@domain.ru",
      "inn": "5544332219",
      "payment_address": "https://domain.ru",
    },
    "correction_info": Object {
      "base_date": "28.11.2017",
      "base_name": "Описание коррекции",
      "base_number": "base_number",
      "type": "instruction",
    },
    "payments": Array [
      Object {
        "sum": 100,
        "type": 2,
      },
    ],
    "vats": Array [
      Object {
        "sum": 18,
        "type": "vat18",
      },
    ],
  },
  "external_id": "correction_id",
  "service": Object {
    "callback_url": "https://domain.ru/callback/new",
  },
  "timestamp": "28.11.2017 23:49:45",
}
`;

exports[`preparePurchaseOrRefund returns prepared data 1`] = `
Object {
  "external_id": "order_id",
  "receipt": Object {
    "client": Object {
      "email": "andreystepanov.blog@gmail.com",
      "phone": "+380939777959",
    },
    "company": Object {
      "email": "support@domain.ru",
      "inn": "5544332219",
      "payment_address": "https://domain.ru",
    },
    "items": Array [
      Object {
        "measurement_unit": "шт.",
        "name": "Товар 1",
        "payment_method": "full_prepayment",
        "payment_object": "service",
        "price": 100,
        "quantity": 2,
        "sum": 200,
        "vat": Object {
          "sum": 36,
          "type": "vat18",
        },
      },
      Object {
        "measurement_unit": "шт.",
        "name": "Товар 2",
        "payment_method": "full_prepayment",
        "payment_object": "service",
        "price": 50,
        "quantity": 1,
        "sum": 50,
        "vat": Object {
          "sum": 9,
          "type": "vat18",
        },
      },
    ],
    "payments": Array [
      Object {
        "sum": 250,
        "type": 1,
      },
    ],
    "total": 250,
  },
  "service": Object {
    "callback_url": "https://domain.ru/callback",
  },
}
`;

exports[`preparePurchaseOrRefund sets custom timestamp and callback_url 1`] = `
Object {
  "external_id": "order_id",
  "receipt": Object {
    "client": Object {
      "email": "andreystepanov.blog@gmail.com",
      "phone": "+380939777959",
    },
    "company": Object {
      "email": "support@domain.ru",
      "inn": "5544332219",
      "payment_address": "https://domain.ru",
    },
    "items": Array [
      Object {
        "measurement_unit": "шт.",
        "name": "Товар 1",
        "payment_method": "full_prepayment",
        "payment_object": "service",
        "price": 100,
        "quantity": 2,
        "sum": 200,
        "vat": Object {
          "sum": 36,
          "type": "vat18",
        },
      },
      Object {
        "measurement_unit": "шт.",
        "name": "Товар 2",
        "payment_method": "full_prepayment",
        "payment_object": "service",
        "price": 50,
        "quantity": 1,
        "sum": 50,
        "vat": Object {
          "sum": 9,
          "type": "vat18",
        },
      },
    ],
    "payments": Array [
      Object {
        "sum": 250,
        "type": 1,
      },
    ],
    "total": 250,
  },
  "service": Object {
    "callback_url": "https://domain.ru/callback/new",
  },
  "timestamp": "28.11.2017 23:49:45",
}
`;

exports[`sending real request to API correction returns validation error 1`] = `"Ошибка валидации входного чека. Ошибочные поля : PropertyRequired: #/external_id; PropertyRequired: #/correction.correction_info.base_number; PropertyRequired: #/correction.correction_info.base_name"`;

exports[`sending real request to API refund returns validation error 1`] = `"Ошибка валидации входного чека. Ошибочные поля : PropertyRequired: #/external_id; PropertyRequired: #/receipt.client; TooFewItems: #/receipt.items"`;

exports[`sending real request to API sell and report returns validation error 1`] = `"Ошибка валидации входного чека. Ошибочные поля : PropertyRequired: #/external_id; PropertyRequired: #/receipt.client; TooFewItems: #/receipt.items"`;
